File: learnLeetCode/Day24/Part4/dna_of_culprit.cpp
#include <bits/stdc++.h>
using namespace std;

// DNA of the Culprit: minimal Hamming distance within threshold k; return index or -1.

int dna_match_index(const string& dna_scene, const vector<string>& candidates, int k){
    int best=-1, bestdiff=INT_MAX; int m=dna_scene.size();
    for(int i=0;i<(int)candidates.size();++i){
        const string &c = candidates[i]; if((int)c.size()!=m) continue;
        int diff=0; for(int j=0;j<m;++j) if(dna_scene[j]!=c[j]) ++diff;
        if(diff<=k && diff<bestdiff){ bestdiff=diff; best=i; }
    }
    return best;
}

int main(){
    string dna_scene="ATGCC"; vector<string> candidates={"ATGCA","ATACC","TTGCC"};
    cout << dna_match_index(dna_scene, candidates, 1) << "\n"; // 0
    return 0;
}

File: Day24/Part4/dna_of_culprit.py
# DNA of the Culprit
# Choose candidate with minimal Hamming distance to dna_scene, requiring diff <= k; return index or -1 if none.
from typing import List

def dna_match_index(dna_scene: str, candidates: List[str], k: int) -> int:
    best_idx = -1
    best_diff = 10**9
    for i, c in enumerate(candidates):
        if len(c) != len(dna_scene):
            continue  # assuming equal length; could extend to sliding match if needed
        diff = sum(1 for a,b in zip(dna_scene, c) if a!=b)
        if diff <= k and diff < best_diff:
            best_diff = diff
            best_idx = i
    return best_idx

if __name__ == "__main__":
    dna_scene = "ATGCC"
    candidates = ["ATGCA", "ATACC", "TTGCC"]
    print(dna_match_index(dna_scene, candidates, 1))  # 0
